internal/endpoints/awsrulesfn/partitions.go

```diff
@@ -0,0 +1,170 @@
+// Code generated by endpoint/awsrulesfn/internal/partition. DO NOT EDIT.
+
+package awsrulesfn
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/aws/smithy-go/ptr"
+)
+
+// Partition provides the metadata describing an AWS partition.
+type Partition struct {
+	Name               string
+	DnsSuffix          string
+	DualStackDnsSuffix string
+	SupportsFIPS       bool
+	SupportsDualStack  bool
+}
+
+// GetPartition returns an AWS [Partition] for the region provided. If the
+// partition cannot be determined nil will be returned.
+func GetPartition(region string) *Partition {
+	if v, ok := regionPartition[region]; ok {
+		p, ok := partitions[v.PartitionID]
+		if !ok {
+			panic(fmt.Sprintf(
+				"bad codegen, unable to get region looked up partition %v from partitions map",
+				v.PartitionID))
+		}
+		p.mergeWith(v.Metadata)
+		return &p
+	}
+
+	for _, r := range regionRegexps {
+		if r.RegionRegex.MatchString(region) {
+			p, ok := partitions[r.PartitionID]
+			if !ok {
+				panic(fmt.Sprintf(
+					"bad codegen, unable to get region regexp looked up partition %v from partitions map",
+					r.PartitionID))
+			}
+			return &p
+		}
+	}
+
+	p, ok := partitions[defaultPartition]
+	if !ok {
+		panic(fmt.Sprintf(
+			"bad codegen, unable to get default partition %v from partitions map",
+			defaultPartition))
+	}
+
+	return &p
+}
+
+func (p *Partition) mergeWith(other partitionMetadata) {
+	if vv := other.Name; vv != nil {
+		p.Name = *vv
+	}
+	if vv := other.DnsSuffix; vv != nil {
+		p.DnsSuffix = *vv
+	}
+	if vv := other.DualStackDnsSuffix; vv != nil {
+		p.DualStackDnsSuffix = *vv
+	}
+	if vv := other.SupportsFIPS; vv != nil {
+		p.SupportsFIPS = *vv
+	}
+	if vv := other.SupportsDualStack; vv != nil {
+		p.SupportsDualStack = *vv
+	}
+}
+
+const defaultPartition = "aws"
+
+var partitions = map[string]Partition{
```

Comment:
**concern/discuss**: It seems like we take the partition data and split it into 3 different views (this partition map of parition id to base config, region partition map, and partition regexes). Why not just generate a list of partitions and have the `GetPartition` function take in a list of partitions?

This would move most of the runtime code for selecting a partition out of codegen and it would be easier to test (e.g. you can feed it a fake list of partitions rather than be tied to what is generated and may change over time). IIRC this is what we do in Kotlin.

Reply:
I agree that the implementation you described in Kotlin seems like a cleaner implementation, I have modified the codegen to only include generation of the partitions json file into a Go data structure